### PR TITLE
Add GRPC Trancoding. Convert Compute.Demo into acceptance tests

### PR DESCRIPTION
**Pull request by viacheslav-rostovtsev**

### CODE REVIEW

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (jskeet) [inline ComputeDemo/ComputeAlphaMainIntegrationTest.cs:97]: Nit: blank line.

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (jskeet) [inline ComputeDemo/ComputeAlphaMainIntegrationTest.cs:103]: I'd suggest computing a deadline at the start of the method (`DateTime.UtcNow + timeout`) and then you can just check `if (DateTime.UtcNow > deadline)`

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (jskeet) [inline ComputeDemo/ComputeAlphaMainIntegrationTest.cs:77]: I can't remember whether we're planning on doing better than this for DIREGAPIC. It would be nice to use the existing `Operation<TResult, TMetadata>` class, but I can see that may not be feasible.

Let's talk about plans when we meet.

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:66]: I'd *suggest* reformatting as this:

```suggestion
        internal List<string> TopLevelFieldNames => _segments
            .Where(segment => segment.TopLevelFieldName != null)
            .Select(s => s.TopLevelFieldName)
            .ToList();
```

Or possibly even put `_segments` on the second line and indent the query calls further. I'm not bothered which.

Basically if a LINQ query is too long for a single line, we tend to prefer "one operator per line".

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:68]: Should this be CreatePropertyPathPatternSegment? I'm not sure how to read "PathPath".

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:107]: I think a comment would be useful here. I'm not really following how this is expected to be used.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:69]: Moving the transcoding switch to later on means we won't find out that the method is misconfigured until we execute it, which I'm not thrilled with. Let's chat about this one.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:116]: Is the period after http meant to be an apostrophe?

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:114]: Nit: I'd move the `=>` to the end of this line.

**Review by jskeet — COMMENTED** [review]
Largely nits (mostly formatting) but a few more significant changes.

We should start thinking (now) about how we're going to integrate this with google-cloud-dotnet when we're a bit further down the line...

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaFinePointsChecks.cs:1]: 👍

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaFinePointsChecks.cs:35]: 👍

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaMainIntegrationTest.cs:1]: 👍

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaMainIntegrationTest.cs:97]: 👍

File: ComputeDemo/ComputeAlphaMainIntegrationTest.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaMainIntegrationTest.cs:103]: 👍

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:66]: 👍

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:68]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:116]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:114]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:128]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:137]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:140]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:152]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:177]: 👍

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (jskeet) [inline ComputeDemo/ComputeAlphaFinePointsChecks.cs:2]: Rumour has it that 2020 is (thankfully) over now.

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (jskeet) [inline ComputeDemo/ComputeAlphaFinePointsChecks.cs:1]: Side-note: it might be worth starting a new project (Google.Cloud.Compute.V1.AcceptanceTests) instead of changing this demo program. I should have suggested that before. (Shouldn't be much work though.)

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:73]: Move into HttpRulePathPatternSegment, like the other factory methods? (The HttpRulePathPatternSegment constructor could then be private, I suspect.)

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:101]: Nit: empty returns comment.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:106]: I suspect this would be simpler (and more efficient) as a foreach loop. I'm happy for it to go in like this and I can create a PR with a suggestion (that can be closed if you prefer this) rather than try to describe exactly what I mean or write a suggestion without validating it.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:83]: Possible optimization: for the majority case where there aren't any query string parameters, skip the call.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:85]: This will ignore the parameters if there's a host.

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (jskeet) [inline Google.Api.Gax.Grpc/Rest/RestMethod.cs:140]: If we use null here instead of creating an empty dictionary on each call, that would be easy to detect in the CreateRequest code.

Alternatively, is there any reason why we need to split this parameter discovery part from the actual append part? Perhaps the TranscodingOutput could just have "path including parameters"?

**Review by jskeet — COMMENTED** [review]
Some more comments to discuss when we meet later :)

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaFinePointsChecks.cs:2]: 👍

File: ComputeDemo/ComputeAlphaFinePointsChecks.cs
Comment (viacheslav-rostovtsev) [reply ComputeDemo/ComputeAlphaFinePointsChecks.cs:1]: 👍

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:73]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:101]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:106]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:83]: 👍

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:85]: 👍 ooh, how did I even miss that

File: Google.Api.Gax.Grpc/Rest/RestMethod.cs
Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:140]: The reason I did it this way is separation of concerns -- I don't want the transcoding part know anything about building the combined URI

Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs:140]: Hmm... given that it already knows about the base URI and the query parameters themselves, I don't think it's much of a violation to include the encoding here. (After all, it knows to encode the body as JSON... URL-encoding the field value doesn't seem overly different to me.)

Comment (viacheslav-rostovtsev) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs]: I think that with the recent changes rollbacking to the contentFactory it's more visible that the transcoding part of the logic doesn't want any more complexity. It can still be done technically. 
(also, I really don't like that transcoding method *knows* that body and query string param names should be JSON but fixin that would invite more ugliness)

Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestMethod.cs]: We can definitely revisit later on.

File: Google.Cloud.Compute.V1.AcceptanceTests/MainIntegrationTest.cs
Comment (jskeet) [inline Google.Cloud.Compute.V1.AcceptanceTests/MainIntegrationTest.cs:49]: (No change needed.) Will remove this when I move the code over - we assume ADC is sufficient, whether that's via a JSON file or not.

File: Google.Cloud.Compute.V1.AcceptanceTests/MainIntegrationTest.cs
Comment (jskeet) [inline Google.Cloud.Compute.V1.AcceptanceTests/MainIntegrationTest.cs:28]: (No change needed). We use a different environment variable (TEST_PROJECT) in google-cloud-dotnet; this will change a little when I move the tests over.

**Review by jskeet — APPROVED** [review]
Various comments for potential future improvements, but nothing that needs to be done before merging.

I won't merge immediately, but if I find that I'd really like to do a GAX build today (before US daytime) I'll merge at that point. Otherwise, feel free to address whichever comments you'd like, rebase into one or two commits, and then merge.

**Comment by jskeet** [comment]
I'm blocked on other things right now, so I'll merge this and do an alpha release.

File: Gax.sln
Comment (viacheslav-rostovtsev) [reply Gax.sln:57]: Sure, I only named it 'Acceptance' because that's what your comment above suggested )

**Pull request merged**
